test/compress/comparing.js

```diff
@@ -222,23 +238,60 @@ issue_2857_5: {
         comparisons: true,
     }
     input: {
-        p && a === undefined || a === null;
-        p && a === undefined || a !== null;
-        p && a !== undefined || a === null;
-        p && a !== undefined || a !== null;
-        p && a === undefined && a === null;
-        p && a === undefined && a !== null;
-        p && a !== undefined && a === null;
-        p && a !== undefined && a !== null;
+        function f(a, p) {
+            p && a === undefined || a === null;
+            p && a === undefined || a !== null;
+            p && a !== undefined || a === null;
+            p && a !== undefined || a !== null;
+            p && a === undefined && a === null;
+            p && a === undefined && a !== null;
+            p && a !== undefined && a === null;
+            p && a !== undefined && a !== null;
+        }
+    }
+    expect: {
+        function f(a, p) {
+            p && void 0 === a || null === a;
+            p && void 0 === a || null !== a;
+            p && void 0 !== a || null === a;
+            p && void 0 !== a || null !== a;
+            p && void 0 === a && null === a;
+            p && void 0 === a && null !== a;
+            p && void 0 !== a && null === a;
+            p && null != a;
+        }
+    }
+}
+
+issue_2857_6: {
+    options = {
+        comparisons: true,
+        pure_getters: "strict",
+        reduce_vars: true,
+    }
+    input: {
+        function f(a) {
+            if (({}).b === undefined || {}.b === null)
+                return a.b !== undefined && a.b !== null;
+        }
+        console.log(f({
+            a: [ null ],
+            get b() {
+                return this.a.shift();
+            }
+        }));
```

Comment:
@Skalman this is a runnable example of https://github.com/mishoo/UglifyJS2/pull/2862#issuecomment-362212330

Reply:
Good to have it as a test!